### PR TITLE
uF Guide Tool & more

### PR DESCRIPTION
**Pull request by MurrayLF**

### CODE REVIEW

**Comment by MurrayLF** [comment]
Wrong branch

File: src/app/view/paperView.js [hunk 1]
```diff
@@ -18,6 +20,10 @@ import * as DXFSolidObjectRenderer from "./render2D/dxfSolidObjectRenderer2D";
 import Layer from "../core/layer";
 import Device from "../core/device";
 import Feature from "../core/feature";
+import Params from "../core/params";
+import Component from "../core/component";
+import MapUtils from "../utils/mapUtils";
+import { components } from "@dagrejs/graphlib/lib/alg";
```
Comment (rkrishnasanka) [inline src/app/view/paperView.js:26]: is this unused ?

File: src/app/view/paperView.js [hunk 2]
```diff
@@ -526,6 +532,141 @@ export default class PaperView {
...
+     * Creates a new component and adds it to viewManager's nonphysicalComponents or the currentDevice's __components
+     * Note: Takes the feature ids as an array
+     * TODO: Modify this to take the MINT String as another parameter
+     * Built for use in uF Guide Tool
+     * @param typeString Type of the Feature
+     * @param params Map of all the paramters
+     * @param featureIDs [String] Feature id's of all the features that will be a part of this component
+     * @param physical Boolean stating whether feature physical or not
+     */
+    addComponent(typeString, paramdata, featureIDs, physical = false) {
```
Comment (rkrishnasanka) [inline src/app/view/paperView.js:644]: Rename this to not be component

File: src/app/view/paperView.js [hunk 3]
```diff
@@ -526,6 +532,141 @@ export default class PaperView {
...
+        const definition = ComponentAPI.getDefinition(typeString);
```
Comment (rkrishnasanka) [inline src/app/view/paperView.js:645]: Let avoid ComponentAPI for all non physical features

File: src/app/view/paperView.js [hunk 4]
```diff
@@ -526,6 +532,141 @@ export default class PaperView {
...
+        for (const i in featureIDs) {
+            newComponent.addFeatureID(featureIDs[i]);
+
+            // Update the component reference
+            feature = this.__viewManagerDelegate.getFeatureByID(featureIDs[i]);
+            feature.referenceID = componentid;
+        }
+
+        if (physical) Registry.currentDevice.addComponent(newComponent);
+        else this.__viewManagerDelegate.addNonphysComponent(newComponent);
```
Comment (rkrishnasanka) [inline src/app/view/paperView.js:666]: Lets not use component

File: src/app/view/tools/renderMouseTool.js [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+    mouseUpHandler(point) {
+        if (this.currentSelectBox) {
+            this.currentSelection = Registry.viewManager.hitFeaturesWithViewElement(this.currentSelectBox);
+            this.selectFeatures();
+        }
+        this.killSelectBox();
+    }
+
+    removeFeatures() {
```
Comment (rkrishnasanka) [inline src/app/view/tools/renderMouseTool.js:61]: Remove boilerplate method here

File: src/app/view/tools/renderMouseTool.js [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            this.currentSelection = [];
+            Registry.canvasManager.render();
+        }
+    }
+
+    mouseDownHandler(event) {
+        const point = MouseTool.getEventPosition(event);
+        const target = this.hitFeature(point);
+        if (target) {
+            if (Registry.viewManager.getComponentForFeatureID(target.featureID) && Registry.viewManager.getComponentForFeatureID(target.featureID).mint == "TEXT") {
```
Comment (rkrishnasanka) [inline src/app/view/tools/renderMouseTool.js:76]: modify logic here to be for non-physical components

File: src/app/view/tools/renderMouseTool.js [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+        }
+    }
+
+    killSelectBox() {
+        if (this.currentSelectBox) {
+            this.currentSelectBox.remove();
+            this.currentSelectBox = null;
+        }
+        this.dragStart = null;
+    }
```
Comment (rkrishnasanka) [inline src/app/view/tools/renderMouseTool.js:111]: Remove boilerplate

File: src/app/view/tools/renderMouseTool.js [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+     * @param paperElement
+     */
+    selectFeature(paperElement) {
+        this.currentSelection.push(paperElement);
+
+        // Find the component that owns this feature and then select all of the friends
+        const component = this.__getComponentWithFeatureID(paperElement.featureID);
+        const connection = this.__getConnectionWithFeatureID(paperElement.featureID);
+        if (component === null && connection === null) {
+            // Does not belong to a component, hence this returns
```
Comment (rkrishnasanka) [inline src/app/view/tools/renderMouseTool.js:129]: This logic will change

File: src/app/view/tools/renderMouseTool.js [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+     * the featureid. Returns null if no component is found.
+     *
+     * @param featureid
+     * @return {Component}
+     * @private
+     */
+    __getComponentWithFeatureID(featureid) {
+        // Get component with the features
+
+        const device_components = Registry.currentDevice.components;
```
Comment (rkrishnasanka) [inline src/app/view/tools/renderMouseTool.js:166]: This logic will need to change once we no longer use components

File: src/app/view/tools/renderMouseTool.js [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+     * @return {*}
+     * @private
+     */
+    __getConnectionWithFeatureID(featureid) {
+        // Get component with the features
+
+        const device_connections = Registry.currentDevice.connections;
+
+        // Check against every component
+        for (const i in device_connections) {
```
Comment (rkrishnasanka) [inline src/app/view/tools/renderMouseTool.js:198]: remove boilderplate

File: src/app/view/tools/renderMouseTool.js [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+     * Function that is fired when we drag and select an area on the paperjs canvas
+     */
+    selectFeatures() {
+        if (this.currentSelection) {
+            for (let i = 0; i < this.currentSelection.length; i++) {
+                const paperFeature = this.currentSelection[i];
+
+                // Find the component that owns this feature and then select all of the friends
+                const component = this.__getComponentWithFeatureID(paperFeature.featureID);
+
```
Comment (rkrishnasanka) [inline src/app/view/tools/renderMouseTool.js:223]: Remove boilerplate

File: src/app/view/tools/renderMouseTool.js [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+        }
+    }
+
+    deselectFeatures() {
+        if (this.rightClickMenu) {
+            this.rightClickMenu.close();
+        }
+        this.paperView.clearSelectedItems();
+        this.currentSelection = [];
+    }
```
Comment (rkrishnasanka) [inline src/app/view/tools/renderMouseTool.js:248]: remove boilerplate

File: src/app/view/tools/renderMouseTool.js [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+
+    rectSelect(point1, point2) {
+        const rect = new paper.Path.Rectangle(point1, point2);
+        rect.fillColor = new paper.Color(0, 0.3, 1, 0.4);
+        rect.strokeColor = new paper.Color(0, 0, 0);
+        rect.strokeWidth = 2;
+        rect.selected = true;
+        return rect;
+    }
```
Comment (rkrishnasanka) [inline src/app/view/tools/renderMouseTool.js:262]: remove boilerplate

File: src/app/view/viewManager.js [hunk 14]
```diff
@@ -72,6 +74,7 @@ export default class ViewManager {
         Registry.currentGrid = this.__grid;
         this.renderLayers = [];
         this.activeRenderLayer = null;
+        this.nonphysComponents = [];
```
Comment (rkrishnasanka) [inline src/app/view/viewManager.js:77]: Add a comment for type

File: src/app/view/viewManager.js [hunk 15]
```diff
@@ -262,6 +265,58 @@ export default class ViewManager {
...
     }
 
+    /**
+     * Adds a component to the device
+     * @param {Component} component Component to be added to the device
+     * @memberof Device
+     * @returns {void}
+     */
+    addNonphysComponent(component) {
+        if (component instanceof Component) {
```
Comment (rkrishnasanka) [inline src/app/view/viewManager.js:275]: modify instanceof check

File: src/app/view/viewManager.js [hunk 16]
```diff
@@ -262,6 +265,58 @@ export default class ViewManager {
...
+        }
+    }
+
+    /**
+     * Returns component object that is identified by the given key
+     * @param {String} key Key to  the component
+     * @return {Component}
+     * @memberof Device
+     */
+    getComponentByID(key) {
```
Comment (rkrishnasanka) [inline src/app/view/viewManager.js:288]: modify name of method

File: src/app/view/viewManager.js [hunk 17]
```diff
@@ -262,6 +265,58 @@ export default class ViewManager {
...
+        //throw new Error("Component with ID " + key + " does not exist");
+    }
+
+    /**
+     * Returns the component identified by the id
+     * @param {string} id ID of the feature to get the component
+     * @return {Component|null}
+     * @memberof Device
+     */
+    getComponentForFeatureID(id) {
```
Comment (rkrishnasanka) [inline src/app/view/viewManager.js:305]: modify method name

File: src/app/view/viewManager.js [hunk 18]
```diff
@@ -1317,7 +1356,8 @@ export default class ViewManager {
      */
     resetToDefaultTool() {
         this.cleanupActiveTools();
-        this.activateTool("MouseSelectTool");
+        //this.activateTool("MouseSelectTool");
+        this.activateTool("RenderMouseTool");
```
Comment (rkrishnasanka) [inline src/app/view/viewManager.js:1360]: Render mouse tool shouldnt be thedefault

File: src/app/view/viewManager.js [hunk 19]
```diff
@@ -1377,6 +1417,7 @@ export default class ViewManager {
      */
     setupTools() {
         this.tools.MouseSelectTool = new MouseSelectTool(this.view);
+        this.tools.RenderMouseTool = new RenderMouseTool(this.view);
```
Comment (rkrishnasanka) [inline src/app/view/viewManager.js:1420]: We need to figure out how to use this/no have this in 3DuF

**Pull request closed without merging**